File: hw02/ekmodel1d.m
%Echebarria B, Karma A.
%Mechanisms for initiation of cardiac discordant alternans.
%The European Physical Journal Special Topics 2007; 146: 217‐31.
%http://link.springer.com/article/10.1140/epjst/e2007-00182-y
%

clear all;

% initial values
num=400;% the number of cells in a cable
v=zeros(num,1);
tmp=zeros(num,1);%temporary array to solve the PDE
h=1.0*ones(num,1);
f=0.9*ones(num,1);
stim=zeros(num,1);

% constants
tauso=15;
taufi=0.8;
tauh1=4.8;
tauh2=10.0;

tausi=4;
tauf1=100;
tauf2=30;

% constants for the PDE
dfu=0.0005;%diffusion coefficient
dx=0.015;%cell size

dt=0.1;% time step (0.1 ms)

pcl=140;%pacing cycle length (ms)
itr=10;% the number of beats
tmax=pcl*itr;%total time

%convert to integer
tnmax=round(tmax/dt);
pcln=round(pcl/dt);
durn=round(1.0/dt);

%array for results
resultv=[];

% main loop
for tn=0:tnmax

  % stimulation
  if (mod(tn,pcln)<durn)
    stim(1:5)=0.2;
  elseif(mod(tn,pcln)==durn)
    stim(1:5)=0;
  end

  minf=((v/0.2).^6)./(1+((v/0.2).^6));
  hinf=1./(1+((v/0.1).^6));
  dinf=((v/0.4).^4)./(1+((v/0.4).^4));
  finf=1./(1+((v/0.1).^4));

  tauh=tauh1+tauh2*exp(-20*((v-0.1).^2));
  tauf=tauf2+(tauf1-tauf2).*v.^3;

  jfi=1 *h.*minf.*(v-1.3)/taufi;
  jsi=1 * f.*dinf.*(v-1.4)/tausi;
  jso=(1-exp(-4*v))/tauso;
  ion=-(jfi+jsi+jso-stim);

  % update variables
  v=v+ion*dt;
  h=h+dt*(hinf-h)./tauh;
  f=f+dt*(finf-f)./tauf;

  %non-flux boundary condition
  
  v(1)=v(3);
  v(num)=v(num-2);
  %solve diffusion equation
  for c=2:num-1
    tmp(c)=v(c)+(v(c-1)+v(c+1)-2*v(c))*dfu*dt/(dx*dx);
    s=v(c);
    s1=v(c-1);
    s2=v(c+1);
    s4 = tmp(c);
  end

  v=tmp;

  if (mod(tn,10)==0)
    resultv=[resultv v];
  end
end

%space-time plot
surf(resultv);
shading interp
axis([0 inf 0 inf]);
view(3)

File: hw02/ekmodel1d.cpp
#include <iostream>
#include <fstream>
#include <cmath>
using namespace std;

int main(void) {
  const int num=400;
  
  //array
  double v[num];//membrane voltage
  double h[num];//Na channel inactivation gate
  double f[num];//Ca channel inactivation gate
  double stim[num];
  double tmp[num];//temporary variable to solve diffusion eqn

  //initial values
  for (int i=0;i<num;i++) {
    v[i]=0;
    h[i]=1.0;
    f[i]=0.9;
    stim[i]=0;
  }

  //constants
  const double dt=0.1;// time step (0.1 ms)
  const double pcl=200;//pacing cycle length (300 ms)
  const int itr=10;//# of beats
  double tmax=pcl*itr;//time to simulate

  int tnmax=tmax/dt;//convert to int
  int pcln=pcl/dt;//convert to int
  int durn=1.0/dt;//duration of stimulation (1.0 ms)

  ofstream os("result.txt");

  //main loop
  for (int tn=0;tn<tnmax;tn++) {
    if (tn%100==0) {//write results every 10 ms
      double t=tn*dt;
      for (int i=0;i<num;i++) {
        os<<v[i]<<"\t";
      }
      os<<endl;
    }

    //stimlate the end of the cable
    if (tn%pcln < durn) {
      for (int i=0;i<5;i++) {
        stim[i]=0.3;
      }
    }
    else {
      for (int i=0;i<5;i++) {
        stim[i]=0;
      }
    }

    //Euler method
    for (int i=0;i<num;i++) {
      const double tauso=15;
      const double taufi=0.8;
      const double tauh1=4.8;
      const double tauh2=10.0;
      const double tausi=4.0;
      const double tauf1=100;
      const double tauf2=30;
      double minf=pow((v[i]/0.2),6)/(1+pow((v[i]/0.2),6));
      double hinf=1/(1+pow((v[i]/0.1),6));
      double dinf=pow((v[i]/0.4),4)/(1+pow((v[i]/0.4),4));
      double finf=1/(1+pow((v[i]/0.1),4));
      double tauh=tauh1+tauh2*exp(-20*pow((v[i]-0.1),2));
      double tauf=tauf2+(tauf1-tauf2)*v[i]*v[i]*v[i];

      double jfi=h[i]*minf*(v[i]-1.3)/taufi;//Fast inward current (Na current)
      double jsi=f[i]*dinf*(v[i]-1.4)/tausi;//Slow inward current (Ca current)
      double jso=(1-exp(-4*v[i]))/tauso;//outward current (K current)
      double ion=-(jfi+jsi+jso-stim[i]);//total transmembrane current

      double dh=(hinf-h[i])/tauh;
      double df=(finf-f[i])/tauf;

      //update variables
      v[i]+=ion*dt;
      h[i]+=dh*dt;
      f[i]+=df*dt;
    }
    //solve Diffusion
    const double dfu=0.0005;//diffusion coefficient
    const double dx=0.015;//0.015 cm
    //non-flux boundary
    v[0]=v[2];
    v[num-1]=v[num-3];
    for (int i=1;i<num-1;i++) {
      tmp[i]=v[i]+(v[i-1]+v[i+1]-2*v[i])*dfu*dt/(dx*dx);
    }
    for (int i=1;i<num-1;i++) {
      v[i]=tmp[i];
    }
  }
  return 0;
}
